### DISCUSSION TITLE
Drag visual implementation approaches

### DISCUSSION THREAD

**Discussion by stevemonaco** [AvaloniaUI/Avalonia#17609] [Q&A]
Avalonia lacks specific APIs to show drag visuals. By "drag visual", I mean content created from the drag source that follows the cursor to indicate what content will be dropped whether as text or a bitmap render. UWP has an API (`DragUI` and [`DragUIOverride`](https://learn.microsoft.com/en-us/uwp/api/windows.ui.xaml.draguioverride?view=winrt-26100)) to support this. See below for an captured example from this [sample](https://github.com/microsoft/Windows-universal-samples/blob/534d40ae64cc6a0ad03b1f0c53236845125abeb2/Samples/XamlDragAndDrop/cs/Scenario2_DragUICustomization.xaml.cs#L86-L94):

![dd-uwpdragvisual2](https://github.com/user-attachments/assets/5ccc13aa-85d4-49e0-b56c-3a7c64021f74)

There are a few aspects I'm concerned about implementing (in the context of a library, not within Avalonia itself):
1. Keeping the drag visual near the cursor. See [this discussion](https://github.com/AvaloniaUI/Avalonia/discussions/14062) for input latency issues.
2. Receiving the event from anywhere within the app and updating the drag visual position. I suppose I could register `DragMove` events with `handledEventsToo` on the Window itself (or all TopLevels)? Currently, I only receive them on top of controls registered for accepting drops.
3. Where to render the drag visual? I'm concerned the adorner layer will suffer input latency. An overlay layer may integrate better with compositor (RequestAnimationFrame), but adds another component that devs need to explicitly integrate into their control tree.

I assume there are some related compositor APIs that might prove useful here, but a good approach is unclear to me.

**Comment by stevemonaco** [comment]
Doing a bit more experimentation and the UWP drag visual works outside of the app's bounds, too, which is a nice bonus.

So that leads me to scenario I forgot: custom cursors that contain the drag visual rendered into the cursor itself. Which would eliminate these issues, but I'm not sure Avalonia would cooperate here as cursors during drag operations appear to wholly depend upon `DragDropEffects` flag before dropping into internal/platform implementations.

**Reply by timunie** [reply]
Hmmm 🤔 

What about a custom cursor? Would that even work? Otherwise I tend to use a tooltip/ popup to follow the pointer. But sure, this will have a lag of 1 or 2 frames

**Reply by mkmatis** [reply]
I believe that this change https://github.com/AvaloniaUI/Avalonia/pull/18287 makes the matter more difficult when using `DragDrop`: pointer move event is now swallowed when dragging, so - as far as I know - it's not possible to update drag visual's position when dragging.

**Reply by stevemonaco** [reply]
Just tested on nightly. You still repeatedly get `DragOver` events. I get about 15/16 per second on a 60fps Windows machine.

**Reply by mkmatis** [reply]
You're right, I switched the event I subscribe to and it works. Thanks!
